### ISSUE TITLE
Switching slaves to read from error

### ISSUE THREAD

**Issue by jonathan-daniel** [riptideio/pymodbus#196]
Hello,

I have two slaves and one master.
When I read from either slaves it works, but when I read from one slave, and then the other I get an error.

DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x1 0x0 0x8 0x15 0xff
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4

Am I missing something?
Thanks

This is the code for slave 2:
```
#!/usr/bin/env python
'''
from pymodbus.server.sync import StartSerialServer

from pymodbus.device import ModbusDeviceIdentification
from pymodbus.datastore import ModbusSequentialDataBlock
from pymodbus.datastore import ModbusSlaveContext, ModbusServerContext

from pymodbus.transaction import ModbusRtuFramer

import logging
logging.basicConfig()
log = logging.getLogger()
log.setLevel(logging.DEBUG)

store = ModbusSlaveContext(
    di = ModbusSequentialDataBlock(0, [17]*100),
    co = ModbusSequentialDataBlock(0, [17]*100),
    hr = ModbusSequentialDataBlock(0, [17]*100),
    ir = ModbusSequentialDataBlock(0, [17]*100))

store2 = ModbusSlaveContext(
    di = ModbusSequentialDataBlock(0, [17]*100),
    co = ModbusSequentialDataBlock(0, [17]*100),
    hr = ModbusSequentialDataBlock(0, [18]*100),
    ir = ModbusSequentialDataBlock(0, [17]*100))

slaves_stores = {
    0x02: store2
}

context = ModbusServerContext(slaves=slaves_stores, single=False)

identity = ModbusDeviceIdentification()
identity.VendorName  = 'Pymodbus'
identity.ProductCode = 'PM'
identity.VendorUrl   = 'http://github.com/riptideio/pymodbus/'
identity.ProductName = 'Pymodbus Server'
identity.ModelName   = 'Pymodbus Server'
identity.MajorMinorRevision = '1.0'

StartSerialServer(context, framer=ModbusRtuFramer, identity=identity, port='/dev/ttyUSB2', stopbits=2, bytesize=8, parity='O', ignore_missing_slaves=True, timeout=.005, baudrate=921600)
```

**Comment by dhoomakethu** [comment]
are you running both the servers on the same port ? Please share the client code as well . Also do mention the version of pymodbus on which this is observed.

**Comment by jonathan-daniel** [comment]
Thanks for the reaction

The master code:

```
#!/usr/bin/env python

from pymodbus.client.sync import ModbusSerialClient as ModbusClient

import time

import logging
logging.basicConfig()
log = logging.getLogger()
log.setLevel(logging.DEBUG)

client = ModbusClient(method='rtu', port='/dev/ttyUSB0', timeout=1, stopbits=1, bytesize=8, parity='N', baudrate=921600)
client.connect()

log.debug("Reading Coils")
rr = client.read_coils(1, 1, unit=0x01)

log.debug("Write to a Coil and read back")
rq = client.write_coil(0, True, unit=1)
rr = client.read_coils(0, 1, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error
assert(rr.bits[0] == True)          # test the expected value

log.debug("Write to multiple coils and read back- test 1")
rq = client.write_coils(1, [True]*8, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error
rr = client.read_coils(1, 21, unit=1)
assert(rr.function_code < 0x80)     # test that we are not an error
resp = [True]*21

resp.extend([False]*3)
assert(rr.bits == resp)         # test the expected value

log.debug("Write to multiple coils and read back - test 2")
rq = client.write_coils(1, [False]*8, unit=1)
rr = client.read_coils(1, 8, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error
assert(rr.bits == [False]*8)         # test the expected value

log.debug("Read discrete inputs")
rr = client.read_discrete_inputs(0, 8, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error

log.debug("Write to a holding register and read back")
rq = client.write_register(1, 10, unit=1)
rr = client.read_holding_registers(1, 1, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error
assert(rr.registers[0] == 10)       # test the expected value

log.debug("Write to multiple holding registers and read back")
rq = client.write_registers(1, [10]*8, unit=1)
rr = client.read_holding_registers(1, 8, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error
assert(rr.registers == [10]*8)      # test the expected value

log.debug("Read input registers")
rr = client.read_input_registers(1, 8, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error

arguments = {
    'read_address':    1,
    'read_count':      8,
    'write_address':   1,
    'write_registers': [20]*8,
}

log.debug("Read write registeres simulataneously")
rq = client.readwrite_registers(unit=1, **arguments)
rr = client.read_holding_registers(1, 8, unit=1)
assert(rq.function_code < 0x80)     # test that we are not an error
assert(rq.registers == [20]*8)      # test the expected value
assert(rr.registers == [20]*8)      # test the expected value

time.sleep(1)

# This is where it crashes, it doesnt matter if i start with 2 or 1, both will work untill switching.

log.debug("======= Read write registeres simulataneously ON THE OTHER SLAVE =====")
rq = client.readwrite_registers(unit=2, **arguments)
rr = client.read_holding_registers(1, 8, unit=2)
assert(rq.function_code < 0x80)     # test that we are not an error
assert(rq.registers == [20]*8)      # test the expected value
assert(rr.registers == [20]*8)      # test the expected value

#---------------------------------------------------------------------------# 
# close the client
#---------------------------------------------------------------------------# 
client.close()
```

One slave is on dev/ttyUSB1 another slave is on dev/ttyUSB2 and the master is on dev/ttyUSB0.
I dont know much about how ports work on Linux, but Im guessing its a different port since its a different dev.

pymodbus version 1.3.1

**Comment by jonathan-daniel** [comment]
So these are probably the most useful logs.

As you can see  It seems slave 2 is interpreting the first byte of the slave 1's response as a function code:

SLAVE 1 
```

DEBUG:pymodbus.server.sync:Started thread to serve client
DEBUG:pymodbus.server.sync:Client Connected [/dev/ttyUSB1:/dev/ttyUSB1]
DEBUG:pymodbus.server.sync:0x1 0x3 0x0 0x0 0x0 0x1 0x84 0xa
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.datastore.context:validate[3] 1:1
DEBUG:pymodbus.datastore.context:getValues[3] 1:1
DEBUG:pymodbus.server.sync:send: 01030200123849
DEBUG:pymodbus.server.sync:0x1 0x3 0x0 0x0 0x0 0x1 0x84 0xa
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.datastore.context:validate[3] 1:1
DEBUG:pymodbus.datastore.context:getValues[3] 1:1
DEBUG:pymodbus.server.sync:send: 01030200123849
DEBUG:pymodbus.server.sync:0x1 0x3 0x0 0x0 0x0 0x1 0x84 0xa
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.datastore.context:validate[3] 1:1
DEBUG:pymodbus.datastore.context:getValues[3] 1:1
DEBUG:pymodbus.server.sync:send: 01030200123849
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.server.sync:requested slave does not exist: 2
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.server.sync:requested slave does not exist: 2
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.server.sync:requested slave does not exist: 2
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.server.sync:requested slave does not exist: 2
```

SLAVE 2

```
DEBUG:pymodbus.server.sync:Started thread to serve client
DEBUG:pymodbus.server.sync:Client Connected [/dev/ttyUSB2:/dev/ttyUSB2]
DEBUG:pymodbus.server.sync:0x1 0x3
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x0 0x0 0x0 0x1 0x84 0xa 0x1 0x3 0x2 0x0 0x12 0x38 0x49
DEBUG:pymodbus.factory:Factory Request[3]
DEBUG:pymodbus.server.sync:requested slave does not exist: 1
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x1 0x3 0x0 0x0 0x0 0x1 0x84 0xa
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x1 0x3 0x2 0x0 0x12 0x38 0x49
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x1 0x3 0x0 0x0 0x0 0x1 0x84 0xa 0x1 0x3 0x2 0x0 0x12 0x38 0x49
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
DEBUG:pymodbus.server.sync:0x2 0x3 0x0 0x0 0x0 0x1 0x84 0x39
DEBUG:pymodbus.factory:Factory Request[1]
ERROR:pymodbus.server.sync:Socket error occurred unpack requires a string argument of length 4
```

**Comment by dhoomakethu** [comment]
I see your serial configuration are different for server and client 

`StartSerialServer(context, framer=ModbusRtuFramer, identity=identity, port='/dev/ttyUSB2', stopbits=2, bytesize=8, parity='O', ignore_missing_slaves=True, timeout=.005, baudrate=921600)`

`client = ModbusClient(method='rtu', port='/dev/ttyUSB0', timeout=1, stopbits=1, bytesize=8, parity='N', baudrate=921600)`

Also I am not sure how /dev/ttyUSB0 is connected to /dev/ttyUSB2 , ideally the master and slave should be on the two ends of same port `MASTER-------/dev/ttyUSBn-----------SLAVE1---SLAVE2` + you are using high baudrate , I would recommend to use the same serial configuration (same stopbits/parity etc ) and try at lower baudrates to see if everything is working as expected.

**Comment by jonathan-daniel** [comment]
My configurations are the same I copied the wrong source, the masters and slave all have each own ports, and I tested various baudrates.

I tried using libmodbus in c and it works, so I know for sure its not the settings or my setup.

**Comment by dhoomakethu** [comment]
If it works with libmodbus, it should work with pymodbus as well.I can not think of any reasons why your code is not working If your serial configurations are same for both master and slave (parity, baudrate etc) . Are you using any real modbus device as your slave? Also if possible please share the correct source code you are using (both python and libmodbus) , just share a link .

**Comment by jonathan-daniel** [comment]
Very weird things going on, it was working the whole day with libmodbus until some point (I did not change anything in the code or hardware). So you might be right its not a problem in the library, I really dont know. I will create a repository soon with all the code I tried. 

To be clear, everything is one device, my laptop, im just using 3 USB ports with three USB to RS485 converters that are connected to each other. 

I can also make a picture of how the A, B, ground and resistors are connected.
But again, it worked the whole day and stopped (afaik) randomly. same symptoms as before.
